src/core/thread/mesh_forwarder.cpp

```diff
@@ -1149,6 +1166,20 @@ ThreadError MeshForwarder::HandleFrameRequest(Mac::Frame &aFrame)
         aFrame.SetFramePending(true);
     }
 
+    aFrame.SetIsARetransmission(mSendMessageIsARetransmission);
+    aFrame.SetMaxTxAttempts(mSendMessageMaxMacTxAttempts);
+
+    if (mSendMessageIsARetransmission && aFrame.GetSecurityEnabled())
```

Comment:
I think we need to break this if-statement out and set the DSN here on retransmission and set the security-related fields if security is enabled.